packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol

```diff
@@ -0,0 +1,136 @@
+    using SafeMath for uint256;
+    using FixedPoint for FixedPoint.Unsigned;
+
+    /**
+     * @notice Swaps required amount of reserve currency to collateral currency which is then used to mint tokens to
+     * liquidate a position within one transaction.
+     * @dev After the liquidation is done the DSProxy that called this method will have an open positon AND pending
+     * liquidation within the financial contract. The bot using the DSProxy should withdraw the liquidation once it has
+     * passed liveness. At this point the position can be manually unwound.
+     * @param uniswapRouter address of the uniswap router used to facilate trades.
+     * @param financialContract address of the financial contract on which the liquidation is occuring.
+     * @param reserveCurrency address of the token to swap for collateral. THis is the common currency held by the DSProxy.
+     * @param liquidatedSponsor address of the sponsor to be liquidated.
+     * @param maxReserverTokenSpent maximum number of reserve tokens to spend in the trade. Bounds slippage.
+     * @param minCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token is below this value.
+     * @param maxCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token exceeds this value.
+     * @param maxTokensToLiquidate max number of tokens to liquidate. For a full liquidation this is the full position debt.
+     * @param deadline abort the trade and liquidation if the transaction is mined after this timestamp.
+     **/
+    function swapMintLiquidate(
+        address uniswapRouter,
+        address financialContract,
+        address reserveCurrency,
+        address liquidatedSponsor,
+        FixedPoint.Unsigned calldata maxReserverTokenSpent,
+        FixedPoint.Unsigned calldata minCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxTokensToLiquidate,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+        FinancialContractInterface fc = FinancialContractInterface(financialContract);
+
+        // 1. Calculate how much collateral needed to mint maxTokensToLiquidate.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
+        FixedPoint.Unsigned memory collateralToMintAtGcr = maxTokensToLiquidate.mul(gcr);
+
+        // 2. Calculate how much collateral is needed for the final fee.
+        FinderInterface finder = FinderInterface(fc.finder());
+        StoreInterface store = StoreInterface(finder.getImplementationAddress("Store"));
+        FixedPoint.Unsigned memory totalCollateralNeeded =
+            collateralToMintAtGcr.add(store.computeFinalFee(fc.collateralCurrency()));
+
+        // 3. Swap reserve currency to get required collateral for the mint + final fee.
+        address[] memory path = new address[](2);
```

Comment:
Does uniswap do any smart routing on-chain? Or does this assume a highly-liquid market between token A and token B, specifically.

Put differently, if there are three pools:
- A <-> C: $100MM in pool
- B <-> C: $100MM in pool
- A <-> B: $100k in pool

Will this go through the A <-> B pool without considering the additional slippage incurred?

This is totally fine/reasonable in v1 (our reserve currency will likely be liquid against almost anything in uniswap, etc etc), but I just wanted to make sure I was understanding this correctly.

Reply:
in this implementation, it assumes a liquid market between reserve and collateral. This might be a bad assumption but for now it's easiest to reason about. I think if we wanted to do better routing we would need to define the router off-chain and provide it to the contract. 

Thinking more about this, it makes sense to keep the reserve currency in ETH so that we always have a liquid pair (95% of liquidity & trades on uniswap happen in eth denominated pairs).